File: FindWeather/FindWeatherMySolution/services/LocationService.cs
using FindWeatherMySolution.config;
using FindWeatherMySolution.interfaces;
using FindWeatherMySolution.models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace FindWeatherMySolution.services
{
    public class LocationService : ILocationService
    {
        private readonly HttpClient _httpClient;
        private readonly ApiKey _apiKey;

        public LocationService()
        {
            _httpClient = new HttpClient();
            _apiKey = new ApiKey();
        }

        public async Task<string> GetIP()
        {
            var ipResponse = await _httpClient.GetStringAsync("https://api.ipify.org");
            var ip = ipResponse.Trim();
            
            Console.WriteLine(ip);

            return ip;
        }

        public async Task<Location> GetLocationByIp()
        {
            try
            {
                var ip = await this.GetIP();
                string ipStackApiKey = _apiKey.GetIpStackApiKey();

                var locationResponse = await _httpClient.GetStringAsync($"http://api.ipstack.com/{ip}?access_key={ipStackApiKey}");

                var locationData = JsonSerializer.Deserialize<JsonElement>(locationResponse);

                return new Location
                {
                    City = locationData.GetProperty("city").GetString(),
                    Latitude = locationData.GetProperty("latitude").GetDouble(),
                    Longitude = locationData.GetProperty("longitude").GetDouble()
                };
            }

            catch (Exception ex)
            {
                Console.WriteLine($"Error getting location: {ex.Message}");
                throw;
            }
        }
    }
}

File: FindWeather/FindWeatherSolution/com.findweather/LocationService.java
package com.findweather;

import com.findweather.models.Location;
import org.json.JSONObject;

public class LocationService {
    private static final String IP_STACK_API_KEY = "b71d442b7e89372c15e566abdffb39eb";

    public Location getLocationByIP() {
        try {
            String ip = APIClient.getPublicIP();
            String apiUrl = "http://api.ipstack.com/" + ip + "?access_key=" + IP_STACK_API_KEY;
            JSONObject response = APIClient.getJSONResponse(apiUrl);
            if (response != null) {
                String city = response.getString("city");
                double lat = response.getDouble("latitude");
                double lon = response.getDouble("longitude");
                System.out.println("Detected Location: " + city);
                return new Location(city, lat, lon);
            }
        } catch (Exception e) {
            System.out.println("Error detecting location: " + e.getMessage());
        }
        return null;
    }
}
